src/string/isomorphism.rs

```diff
@@ -0,0 +1,56 @@
+/*
+Given two strings s and t, determine whether they are isomorphic.
+The two strings are isomorphic if the characters in s can be replaced by some mapping relation to get t
+*/
+use std::collections::HashMap;
+
+pub fn is_isomorphic(s: &str, t: &str) -> bool {
+    if s.len() != t.len() {
+        return false;
+    }
+    let sv: Vec<char> = s.chars().collect();
+    let tv: Vec<char> = t.chars().collect();
+    let mut sr1 = String::new();
+    let mut map1 = HashMap::new();
+    for i in 0..sv.len() {
+        if let Some(x) = map1.get(&sv[i]) {
+            sr1.push(*x);
+        } else {
+            map1.insert(sv[i], tv[i]);
+            sr1.push(tv[i]);
+        }
+    }
+    let mut sr2 = String::new();
+    let mut map2 = HashMap::new();
+    for i in 0..sv.len() {
+        if let Some(x) = map2.get(&tv[i]) {
+            sr2.push(*x);
+        } else {
+            map2.insert(tv[i], sv[i]);
+            sr2.push(sv[i]);
+        }
+    }
+    sr1 == t && sr2 == s
+}
+
+#[cfg(test)]
+mod tests {
+    use super::is_isomorphic;
+
+    #[test]
+    fn test_is_isomorphic1() {
+        assert!(is_isomorphic("egg", "add"));
+    }
+    #[test]
+    fn test_is_isomorphic2() {
+        assert!(!is_isomorphic("egg", "adc"));
+    }
+    #[test]
+    fn test_unicode_is_isomorphic() {
+        assert!(is_isomorphic("天苍苍", "野茫茫"));
+    }
+    #[test]
+    fn test_empty_is_isomorphic() {
+        assert!(is_isomorphic("", ""));
+    }
+}
```

Comment:
What do you think about reducing _code duplication_, adding some new test cases and _checks_?

```suggestion
#[cfg(test)]
mod tests {
    use super::is_isomorphic;

    macro_rules! test_is_isomorphic {
        ($($name:ident: $inputs:expr,)*) => {
        $(
            #[test]
            fn $name() {
                let (s, t, expected) = $inputs;
                assert_eq!(is_isomorphic(s, t), expected);
                assert_eq!(is_isomorphic(t, s), expected);
                assert!(is_isomorphic(s, s));
                assert!(is_isomorphic(t, t));
            }
        )*
        }
    }

    test_is_isomorphic! {
        isomorphic: ("egg", "add", true),
        not_isomorphic: ("egg", "adc", false),
        isomorphic_unicode: ("天苍苍", "野茫茫", true),
        isomorphic_unicode_different_byte_size: ("abb", "野茫茫", true),
        empty: ("", "", true),
        different_length: ("abc", "abcd", false),
    }
}
```

Reply:
The unit tests will be written later using macros